0000-ilp-ping-protocol/0000-ilp-ping-protocol.md

```diff
@@ -0,0 +1,78 @@
+   - **Amount**: Any amount chosen by the Initiator.
+   - **Expiry**: Any appropriate expiry.
+   - **Condition**: The Known Condition, as defined in [Terminology](#terminology).
+   
+2. Upon receiving the Prepare packet, the Recipient MUST fulfill the payment using the following information:
+   - **Fulfillment**: The Known Fulfillment, as defined above.
+
+Note that a Recipient _MAY_ reject the payment if appropriate, for example due to an insufficient amount, invalid expiry, or for other factors.
+
+## Bidirectional Mode
+
+1. Initiator generates a new random 32-byte value to use as a Fulfilment (`F`) and generates the corresponding condition (`C`).
+2. Initiator sends an ILP Prepare packet to the Recipient with the following details:
+   - **Destination**: The Recipient's Interledger address.
+   - **Amount**: Any amount chosen by the Initiator.
+   - **Expiry**: An appropriate expiry that is long enough to allow for two round trips.
+   - **Condition**: The condition `C`.
+   - **Data**: A concatenated series of bytes with the following information:
+      - the bytes of the ASCII string `ECHOECHOECHOECHO`. In binary this value is `0x4543484F4543484F4543484F4543484F`.
+      - the byte `0x00`
+      - The ILP address of the Initiator (i.e., the return address) as an OER-encoded, variable length IA5 string.
+4. Upon receiving the Prepare packet, the Recipient identifies that it is a bidirectional Ping request by confirming the following:
+  1. The packet is addressed directly to itself (i.e. it has no tx or child suffix).
+  1. The packet has a payload conforming to the specification above (e.g., the byte following the `ECHOECHOECHOECHO` prefix in the payload is `0x00`).
+5. The Recipient DOES NOT immediately send an ILP Fulfill response. Instead the Recipient sends a new ILP `Prepare` packet addressed to the address parsed from the original original packet's payload. This new packet has following details:original 
+    - *Destination*: The Interledger address found in the original packet's data payload.
+    - *Amount*: Any amount chosen by the Initiator.
+    - *Expiry*: An expiry that is marginally smaller than the expiry on the original ILP Prepare.
+    - *Condition*: The same condition `C`.
+    - *Data*: A concatenated series of bytes with the following information:
+      - the bytes of the ASCII string `ECHOECHOECHOECHO` (i.e., `0x4543484F4543484F4543484F4543484F`).
+      - the byte `0x01`
+7. Upon receiving this packet, the Initiator identifies that it is a Pong by the fact that it is addressed to the address used in the original Ping request.
+8. The Initiator fulfills this second packet using the fulfilment `F`.
+9. Upon receiving the ILP `Fulfill` packet from the Initiator, the Recipient then fulfills the original Ping request using the same fulfillment `F`.
+
+## Implementation Recommendations
+
+1. The Initiator SHOULD choose a unique address that is only used for a single particular pinging session.
```

Comment:
I think you mean that they should append a unique suffix to their normal ILP address. If they choose a random one, the packet won't actually reach them.

Reply:
Agree. I've updated this verbiage -- though see my comment above. We need to decide if the protocol should be triggered on ILP addresses that _start with_ the Connector address, or instead  _equal_ the Connector address.

Reply:
I think it's easiest for the initiator to identify ping responses if they use a special suffix appended to their ILP address